### PR TITLE
Constraint tabs

### PR DESCRIPTION
**Pull request by u-train**
Adds constraints to UI library. Controllers are built upon the tab constraints.

### CODE REVIEW

**Comment by Sanjay-B** [comment]
I won't be reviewing this without proper documentation. 
``source/libraries/UI/constraints/controller.lua``

File: source/application/main.lua [hunk 1]
```diff
@@ -8,4 +8,32 @@ require("tevgit:source/application/utilities/debug/output.lua")
 require("tevgit:source/application/utilities/debug/keybinds.lua")
 
 -- List of stuff that updates
-require("tevgit:source/application/updater/main.lua")
\ No newline at end of file
+--require("tevgit:source/application/updater/main.lua")
```
Comment (Sanjay-B) [inline source/application/main.lua:11]: This file is not meant to have code. It's an entry point file where it also points to other main files. It should not have your test code. Test code should never make its way into production. Clean this up.

File: source/libraries/UI/constraints/controllers/gridLayout.lua [hunk 2]
```diff
@@ -0,0 +1,5 @@
+local newTabResolver = require("tevgit:source/libraries/UI/constraints/controllers/verticalNavLayout.lua")
+
+return function()
```
Comment (Sanjay-B) [inline source/libraries/UI/constraints/controllers/gridLayout.lua:3]: I'm confused as to what this is for. I'm assuming that this is still test a work in progress. You're requiring a module & returning an empty function. I need more context.

File: source/libraries/UI/constraints/controllers/horizontialLayoutAuto.lua [hunk 3]
```diff
@@ -0,0 +1,117 @@
+local count = function(x)
```
Comment (Sanjay-B) [inline source/libraries/UI/constraints/controllers/horizontialLayoutAuto.lua:1]: Document this. I know what it does & how it works but, if we don't document every function, our codebase will turn into a mess. Correct this.

File: source/libraries/UI/constraints/controllers/horizontialLayoutAuto.lua [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+	end
+	return c
+end
+
+return function()
+	--[[
+		@description
+			Creates a new horizontialLayoutAuto that automatically fits itself.
+			So, that means there is no minimum or maximum really.
+
```
Comment (Sanjay-B) [inline source/libraries/UI/constraints/controllers/horizontialLayoutAuto.lua:14]: Missing documentation ``@parameter`` tag. Fix this.

File: source/libraries/UI/constraints/controllers/horizontialLayoutAuto.lua [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+	local public = {}
+	local objects = {}
+
+	public.container = core.construct("guiFrame")
+
+	public.refresh = function()
+		--[[
+			@description
+				Refreshes the container.
```
Comment (Sanjay-B) [inline source/libraries/UI/constraints/controllers/horizontialLayoutAuto.lua:27]: Missing ``@parameter`` & ``@return`` tags. Fix this.

File: source/libraries/UI/constraints/controllers/horizontialLayoutAuto.lua [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+	public.addObject = function(tag, object)
+		--[[
+			@description
+				Add object
+
+			@parameter
+				string, tag
+					Tag.
+				guiObject, object
+					Object.
```
Comment (Sanjay-B) [inline source/libraries/UI/constraints/controllers/horizontialLayoutAuto.lua:78]: Missing ``@return`` documentation tag. Correct this.

File: source/libraries/UI/constraints/controllers/horizontialLayoutAuto.lua [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+				guiObject, object
+		]]
+
+		local object = objects[tag]
+		objects[tag] = nil
+		return object
+	end
+
+	return public
+end;
```
Comment (Sanjay-B) [inline source/libraries/UI/constraints/controllers/horizontialLayoutAuto.lua:117]: Remove the ``;``. I shouldn't see ``;`` in your code.

File: source/libraries/UI/constraints/controllers/verticalLayoutAuto.lua [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+	end
+	return c
+end
+
+return function()
+	--[[
+		@description
+			Creates a new verticalLayoutAuto that automatically fits itself.
+			So, that means there is no minimum or maximum really.
+
```
Comment (Sanjay-B) [inline source/libraries/UI/constraints/controllers/verticalLayoutAuto.lua:14]: Missing ``@parameter`` documentation tag. Fix this.

File: source/libraries/UI/constraints/controllers/verticalLayoutAuto.lua [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+	local public = {}
+	local objects = {}
+
+	public.container = core.construct("guiFrame")
+
+	public.refresh = function()
+		--[[
+			@description
+				Refreshes the container.
+		]]
```
Comment (Sanjay-B) [inline source/libraries/UI/constraints/controllers/verticalLayoutAuto.lua:28]: Missing ``@parameter`` & ``@return`` documentation tags. Fix this.

File: source/libraries/UI/constraints/controllers/verticalLayoutAuto.lua [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+			
+			@return
+				table, [children]
+		]]
+
+		local children = public.container.children
+		
+		if destroyChildren then
+			object:destroyChildren()
+			object:destroy()
```
Comment (Sanjay-B) [inline source/libraries/UI/constraints/controllers/verticalLayoutAuto.lua:58]: If you're destroying the object's children & then the object, just destroy the object & kill two birds with one stone.

File: source/libraries/UI/constraints/controllers/verticalLayoutAuto.lua [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+	public.addObject = function(tag, object)
+		--[[
+			@description
+				Add object
+
+			@parameter
+				string, tag
+					Tag.
+				guiObject, object
+					Object.
```
Comment (Sanjay-B) [inline source/libraries/UI/constraints/controllers/verticalLayoutAuto.lua:78]: Missing ``@return`` documentation tag. Fix it.

File: source/libraries/UI/constraints/controllers/verticalLayoutAuto.lua [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+				guiObject, object
+		]]
+		
+		local object = objects[tag]
+		objects[tag] = nil
+		return object
+	end
+
+	return public
+end;
```
Comment (Sanjay-B) [inline source/libraries/UI/constraints/controllers/verticalLayoutAuto.lua:117]: Remove ``;``. I shouldn't see ``;`` in your code.

File: source/libraries/UI/constraints/controllers/verticalNavLayout.lua [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+	local public = {}
+
+	public.offset = 5
+	public.container = core.construct("guiFrame")
+
+	public.refresh = function()
+		--[[
+			@description
+				Refreshes the container.
+		]]
```
Comment (Sanjay-B) [inline source/libraries/UI/constraints/controllers/verticalNavLayout.lua:14]: Missing ``@parameter`` & ``@return`` documentation tags. Fix it.

File: source/libraries/UI/constraints/controllers/verticalNavLayout.lua [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+			
+			@return
+				table, [children]
+		]]
+
+		local children = public.container.children
+		
+		if destroyChildren then
+			object:destroyChildren()
+			object:destroy()
```
Comment (Sanjay-B) [inline source/libraries/UI/constraints/controllers/verticalNavLayout.lua:44]: Refer to the other comment made about this.

File: source/libraries/UI/constraints/controllers/verticalNavLayout.lua [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+	public.secondaryObject = nil
+	public.min = 0
+	public.max = 50
+
+	public.navBar = createNavBar()
+
+	public.refresh = function()
+		--[[
+			@description
+				Refreshes the layout.
```
Comment (Sanjay-B) [inline source/libraries/UI/constraints/controllers/verticalNavLayout.lua:76]: Missing ``@parameter`` & ``@return`` documentation tags. Correct this.

File: source/libraries/UI/constraints/controllers/verticalNavLayout.lua [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+			
+			@return
+				table, [children]
+		--]]
+
+		local children = public.container.children
+		
+		if destroyChildren then
+			object:destroyChildren()
+			object:destroy()
```
Comment (Sanjay-B) [inline source/libraries/UI/constraints/controllers/verticalNavLayout.lua:106]: Refer to the last two comments made about this.

File: source/libraries/UI/constraints/tabResolver.lua [hunk 17]
```diff
@@ -0,0 +1,260 @@
...
+		table.insert(currentTabs, relativeTab)
+		return getRelativeTabs(relativeTab, currentTabs)
+	end
+
+	return currentTabs
+end
+
+local new = function()
+	local public = {}
+	public.tabs = { x = {}; y = {} }
```
Comment (Sanjay-B) [inline source/libraries/UI/constraints/tabResolver.lua:56]: Remove ``;`` & replace with a ``,``. I should never see a ``;``.

File: source/libraries/UI/constraints/tabResolver.lua [hunk 18]
```diff
@@ -0,0 +1,260 @@
...
+				tab, tab
+		]]
+
+		max = max or min
+		pref = pref or max/2
+		weight = weight or 1
+
+		public.tabs[axis][tag] =
+		{
+			min = min;
```
Comment (Sanjay-B) [inline source/libraries/UI/constraints/tabResolver.lua:87]: Remove ``;`` & replace with a ``,``. I should never see a ``;``.

File: source/libraries/UI/constraints/tabResolver.lua [hunk 19]
```diff
@@ -0,0 +1,260 @@
...
+			max = max;
```
Comment (Sanjay-B) [inline source/libraries/UI/constraints/tabResolver.lua:88]: Remove ``;`` & replace with a ``,``. I should never see a ``;``.

File: source/libraries/UI/constraints/tabResolver.lua [hunk 20]
```diff
@@ -0,0 +1,260 @@
...
+			weight = weight;
```
Comment (Sanjay-B) [inline source/libraries/UI/constraints/tabResolver.lua:89]: Remove ``;`` & replace with a ``,``. I should never see a ``;``.

File: source/libraries/UI/constraints/tabResolver.lua [hunk 21]
```diff
@@ -0,0 +1,260 @@
...
+
+			tag = tag;
```
Comment (Sanjay-B) [inline source/libraries/UI/constraints/tabResolver.lua:91]: Remove ``;``. I should never see a ``;``.

File: source/libraries/UI/constraints/tabResolver.lua [hunk 22]
```diff
@@ -0,0 +1,260 @@
...
+			integer, offset
+				The offset away from the border.
+
+			@returns
+				tab, tab
+		]]
+
+		public.tabs[axis][tag] =
+		{
+			tag = tag;
```
Comment (Sanjay-B) [inline source/libraries/UI/constraints/tabResolver.lua:126]: Remove ``;`` & replace with a ``,``. I should never see a ``;``.

File: source/libraries/UI/constraints/tabResolver.lua [hunk 23]
```diff
@@ -0,0 +1,260 @@
...
+			offset = offset;
```
Comment (Sanjay-B) [inline source/libraries/UI/constraints/tabResolver.lua:127]: Remove ``;`` & replace with a ``,``. I should never see a ``;``.

File: source/libraries/UI/constraints/tabResolver.lua [hunk 24]
```diff
@@ -0,0 +1,260 @@
...
+			relativeTabs = {};
```
Comment (Sanjay-B) [inline source/libraries/UI/constraints/tabResolver.lua:128]: Remove ``;``. I should never see a ``;``.

**Review by Sanjay-B — CHANGES_REQUESTED** [review]

File: source/libraries/UI/constraints/controllers/gridLayout.lua (hunk 2 above)
Comment (u-train) [reply source/libraries/UI/constraints/controllers/gridLayout.lua:3]: Unimplemented.

File: source/libraries/UI/constraints/tabResolver.lua (hunk 20 above)
Comment (u-train) [reply source/libraries/UI/constraints/tabResolver.lua:89]: Why must you be attacking the colon character?

File: source/libraries/UI/constraints/controllers/gridLayout.lua (hunk 2 above)
Comment (Sanjay-B) [reply source/libraries/UI/constraints/controllers/gridLayout.lua:3]: We need this. Implement it.

File: source/libraries/UI/constraints/tabResolver.lua (hunk 20 above)
Comment (Sanjay-B) [reply source/libraries/UI/constraints/tabResolver.lua:88]: This Lua & not Java. ``;`` don't match the majority of the codebase. Looking for consistency not personal preference.

**Review by Sanjay-B — APPROVED** [review]

**Pull request merged**